NPE after deleting a role

## Gentics Mesh Version, operating system, or hardware.
- v1.4.0 and v1.4.3

## Operating System
- Linux using Docker-Compose

## Problem
NPE when indexing after deleting a role.

## Reproducer
1. create and remove role
2. start reindex

## Stacktrace
```
docker-compose[3564]: mesh-server      | Caused by: java.lang.NullPointerException: null
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.search.index.AbstractTransformer.addPermissionInfo(AbstractTransformer.java:123)
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.search.index.project.ProjectTransformer.toDocument(ProjectTransformer.java:34)
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.search.index.project.ProjectTransformer.toDocument(ProjectTransformer.java:17)
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.search.index.entry.AbstractIndexHandler.lambda$null$2(AbstractIndexHandler.java:222)
docker-compose[3564]: mesh-server      |         at com.gentics.madl.tx.TxFactory.lambda$tx$1(TxFactory.java:71)
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.graphdb.OrientDBDatabase.tx(OrientDBDatabase.java:399)
docker-compose[3564]: mesh-server      |         at com.gentics.madl.tx.TxFactory.tx(TxFactory.java:70)
docker-compose[3564]: mesh-server      |         at com.gentics.mesh.search.index.entry.AbstractIndexHandler.lambda$null$3(AbstractIndexHandler.java:222)
docker-compose[3564]: mesh-server      |         at io.reactivex.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:63)
docker-compose[3564]: mesh-server      |         ... 62 common frames omitted
```

--- a/CHANGELOG.adoc
+++ b/CHANGELOG.adoc
@@ -25,6 +25,11 @@ include::content/docs/variables.adoc-include[]
 
 * The Mesh Server will require Java 11 with the release of 2.0.0. The runtime support for Java 8 will be dropped. The Mesh Java REST client will still be usable with Java 8.
 
+[[v1.4.7]]
+== 1.4.7 (TBD)
+
+icon:checked[] Elasticsearch: Fixed a bug that prevented Elasticsearch synchronization of projects after deleting a role. link:https://github.com/gentics/mesh/issues/1041[#1041]
+
 [[v1.4.6]]
 == 1.4.6 (20.04.2020)
 

--- a/core/src/main/java/com/gentics/mesh/core/data/generic/PermissionProperties.java
+++ b/core/src/main/java/com/gentics/mesh/core/data/generic/PermissionProperties.java
@@ -3,6 +3,7 @@ package com.gentics.mesh.core.data.generic;
 import static com.gentics.mesh.core.data.relationship.GraphPermission.READ_PERM;
 import static org.apache.commons.lang3.StringUtils.isEmpty;
 
+import java.util.Objects;
 import java.util.Set;
 import java.util.stream.Stream;
 
@@ -34,7 +35,10 @@ public class PermissionProperties {
 			? Stream.empty()
 			: roleUuids.stream();
 		RoleRoot roleRoot = boot.roleRoot();
-		return new TraversalResult<>(stream.map(roleRoot::findByUuid));
+		return new TraversalResult<>(stream
+			.map(roleRoot::findByUuid)
+			.filter(Objects::nonNull)
+		);
 	}
 
 	public PermissionInfo getRolePermissions(MeshVertex vertex, InternalActionContext ac, String roleUuid) {

--- a/core/src/test/java/com/gentics/mesh/search/RoleDeleteSyncTest.java
+++ b/core/src/test/java/com/gentics/mesh/search/RoleDeleteSyncTest.java
@@ -1,9 +1,14 @@
 package com.gentics.mesh.search;
 
+import static com.gentics.mesh.test.TestDataProvider.PROJECT_NAME;
+import static com.gentics.mesh.test.context.MeshTestHelper.getSimpleQuery;
+import static org.assertj.core.api.Assertions.assertThat;
+
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
+import com.gentics.mesh.core.rest.project.ProjectListResponse;
 import com.gentics.mesh.core.rest.project.ProjectResponse;
 import com.gentics.mesh.core.rest.role.RolePermissionRequest;
 import com.gentics.mesh.core.rest.role.RoleResponse;
@@ -23,6 +28,8 @@ public class RoleDeleteSyncTest extends AbstractMultiESTest {
 
 	@Test
 	public void testSyncAfterDeletedRole() throws Exception {
+		recreateIndices();
+
 		RoleResponse role = createRole("testRole");
 		ProjectResponse project = getProject();
 		client().updateRolePermissions(
@@ -33,5 +40,7 @@ public class RoleDeleteSyncTest extends AbstractMultiESTest {
 		client().deleteRole(role.getUuid()).blockingAwait();
 
 		recreateIndices();
+		ProjectListResponse response = client().searchProjects(getSimpleQuery("name", PROJECT_NAME)).blockingGet();
+		assertThat(response.getData()).isNotEmpty();
 	}
 }

--- a/core/src/test/java/com/gentics/mesh/search/RoleDeleteSyncTest.java
+++ b/core/src/test/java/com/gentics/mesh/search/RoleDeleteSyncTest.java
@@ -1,0 +1,37 @@
+package com.gentics.mesh.search;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import com.gentics.mesh.core.rest.project.ProjectResponse;
+import com.gentics.mesh.core.rest.role.RolePermissionRequest;
+import com.gentics.mesh.core.rest.role.RoleResponse;
+import com.gentics.mesh.test.TestSize;
+import com.gentics.mesh.test.context.ElasticsearchTestMode;
+import com.gentics.mesh.test.context.MeshTestSetting;
+
+/**
+ * Reproducer for https://github.com/gentics/mesh/issues/1041
+ */
+@RunWith(Parameterized.class)
+@MeshTestSetting(startServer = true, testSize = TestSize.PROJECT)
+public class RoleDeleteSyncTest extends AbstractMultiESTest {
+	public RoleDeleteSyncTest(ElasticsearchTestMode elasticsearch) throws Exception {
+		super(elasticsearch);
+	}
+
+	@Test
+	public void testSyncAfterDeletedRole() throws Exception {
+		RoleResponse role = createRole("testRole");
+		ProjectResponse project = getProject();
+		client().updateRolePermissions(
+			role.getUuid(),
+			"/projects/" + project.getUuid(),
+			RolePermissionRequest.grantAll()
+		).blockingAwait();
+		client().deleteRole(role.getUuid()).blockingAwait();
+
+		recreateIndices();
+	}
+}

--- a/core/src/test/java/com/gentics/mesh/test/context/TestHelper.java
+++ b/core/src/test/java/com/gentics/mesh/test/context/TestHelper.java
@@ -300,10 +300,14 @@ public interface TestHelper extends EventHelper, ClientHelper {
 		call(() -> client().deleteGroup(uuid));
 	}
 
-	default public RoleResponse createRole(String roleName, String groupUuid) {
+	default RoleResponse createRole(String roleName) {
 		RoleCreateRequest roleCreateRequest = new RoleCreateRequest();
 		roleCreateRequest.setName(roleName);
-		RoleResponse roleResponse = call(() -> client().createRole(roleCreateRequest));
+		return call(() -> client().createRole(roleCreateRequest));
+	}
+
+	default public RoleResponse createRole(String roleName, String groupUuid) {
+		RoleResponse roleResponse = createRole(roleName);
 		client().addRoleToGroup(groupUuid, roleResponse.getUuid()).blockingAwait();
 		return roleResponse;
 	}

--- a/rest-model/src/main/java/com/gentics/mesh/core/rest/role/RolePermissionRequest.java
+++ b/rest-model/src/main/java/com/gentics/mesh/core/rest/role/RolePermissionRequest.java
@@ -25,7 +25,7 @@ public class RolePermissionRequest implements RestModel {
 
 	/**
 	 * Return a set of permissions that should be set to the affected elements.
-	 * 
+	 *
 	 * @return
 	 */
 	public PermissionInfo getPermissions() {
@@ -34,7 +34,7 @@ public class RolePermissionRequest implements RestModel {
 
 	/**
 	 * Flag that indicated that the request should be executed recursively.
-	 * 
+	 *
 	 * @return Flag value
 	 */
 	public Boolean getRecursive() {
@@ -43,7 +43,7 @@ public class RolePermissionRequest implements RestModel {
 
 	/**
 	 * Set the flag which indicated whether the permission changes should be applied recursively.
-	 * 
+	 *
 	 * @param recursive
 	 *            Recursive flag value
 	 * @return Fluent API
@@ -67,5 +67,13 @@ public class RolePermissionRequest implements RestModel {
 		}
 		permissionsInfo.setOthers(false);
 		return rolePermissionRequest;
+	}
+
+	/**
+	 * Creates a {@link RolePermissionRequest} that is non-recursive and grants all permissions.
+	 * @return
+	 */
+	public static RolePermissionRequest grantAll() {
+		return withPermissions(Permission.values());
 	}
 }

